hudi-client/hudi-client-common/src/main/java/org/apache/hudi/table/upgrade/BaseTwoToOneDowngradeHandler.java

```diff
@@ -0,0 +1,140 @@
+      try {
+        convertToDirectMarkers(
+            commitInstant.getTimestamp(), table, context, config.getMarkersDeleteParallelism());
+      } catch (IOException e) {
+        throw new HoodieException("Converting marker files to DIRECT style failed during downgrade", e);
+      }
+    }
+    return Collections.EMPTY_MAP;
+  }
+
+  abstract HoodieTable getTable(HoodieWriteConfig config, HoodieEngineContext context);
+
+  /**
+   * Converts the markers in new format(timeline server based) to old format of direct markers,
+   * i.e., one marker file per data file, without MARKERS.type file.
+   * This needs to be idempotent.
+   * 1. read all markers from timeline server based marker files
+   * 2. create direct style markers
+   * 3. delete marker type file
+   * 4. delete timeline server based marker files
+   *
+   * @param commitInstantTime instant of interest for marker conversion.
+   * @param table             instance of {@link HoodieTable} to use
+   * @param context           instance of {@link HoodieEngineContext} to use
+   * @param parallelism       parallelism to use
+   */
+  private void convertToDirectMarkers(final String commitInstantTime,
+                                      HoodieTable table,
+                                      HoodieEngineContext context,
+                                      int parallelism) throws IOException {
+    String markerDir = table.getMetaClient().getMarkerFolderPath(commitInstantTime);
+    FileSystem fileSystem = FSUtils.getFs(markerDir, context.getHadoopConf().newCopy());
+    Option<MarkerType> markerTypeOption = MarkerUtils.readMarkerType(fileSystem, markerDir);
+    if (markerTypeOption.isPresent()) {
+      switch (markerTypeOption.get()) {
+        case TIMELINE_SERVER_BASED:
+          // Reads all markers written by the timeline server
+          Map<String, Set<String>> markersMap =
+              MarkerUtils.readTimelineServerBasedMarkersFromFileSystem(
+                  markerDir, fileSystem, context, parallelism);
+          DirectWriteMarkers directWriteMarkers = new DirectWriteMarkers(table, commitInstantTime);
+          // Recreates the markers in the direct format
+          for (Set<String> markers : markersMap.values()) {
+            markers.forEach(directWriteMarkers::create);
+          }
+          // Delete marker type file
+          MarkerUtils.deleteMarkerTypeFile(fileSystem, markerDir);
+          // delete timeline server based markers
+          deleteTimelineBasedMarkerFiles(markerDir, fileSystem);
+          break;
+        default:
+          throw new HoodieException("The marker type \"" + markerTypeOption.get().name() + "\" is not supported.");
+      }
+    } else {
+      // incase of partial failures during downgrade, there is a chance that marker type file was deleted, but timeline server based marker files are left.
+      // so delete them if any
+      deleteTimelineBasedMarkerFiles(markerDir, fileSystem);
```

Comment:
this is okay, since only deletes the files with the prefix. so will not touch anything written using direct markers

Reply:
Yes.  I also changed the filter to be strict to avoid accidental matching.